### PR TITLE
Masked softmax (& log-softmax)

### PR DESCRIPTION
**Pull request by MoisesHer**
## Description ##
Masked softmax and log-softmax operators: they mask out values not included in the mask 

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] Two new operators: masked_softmax and log_masked_softmax
- [x] Forward and Backward passes included
- [x] Optionally, an scaling factor _s_ can be passed as parameter. Each value _x_ in the tensor will be converted into x/s
- [x] Normalization by Max value (activated by default) can be deactivated. This can improve performance when no required
- [x] Automatic broadcast of mask for axis with size=1 (requires same number of dimensions as input data)
- [x] CPU implementation is included

@sxjscience

### CODE REVIEW

**Comment by MoisesHer** [comment]
I think mask should have same shape as input for having a flexible and generic operator.
If any broadcast should happens, I think should be managed externally.
But, I am open mind if people think this op need to deal with brodcasting.

**Comment by sxjscience** [comment]
@MoisesHer Currently, the design of GluonNLP attention cell assumes that the mask can be broadcastable. Is it possible to just add the broadcasting support to the head-axis? https://github.com/dmlc/gluon-nlp/blob/20321598a732d254d4a9e036eecbe0ae8abb48a0/src/gluonnlp/attention_cell.py#L249

File: src/operator/nn/softmax-inl.h [hunk 1]
```diff
@@ -161,6 +161,23 @@ inline void Softmax(Stream<cpu> *s, DType *in, OType *out, IType *length,
   }
 }
 
+template<typename OP, bool negate, typename AType, typename DType,
+         typename OType, typename IType, int ndim>
+inline void MaskedSoftmax(Stream<cpu> *s, DType *in, OType *out, IType *mask,
+                          Shape<ndim> shape, int axis, const double scale,
+                          const double temperature, bool normalize) {
+  common::LogOnce("masked_softmax is not supported on CPU");
```
Comment (sxjscience) [inline src/operator/nn/softmax-inl.h:169]: We need to support CPU. We may try to have a very naive implementation.

**Comment by sxjscience** [comment]
We can try to revise the implementation of GluonNLP to see if the new OP can pass the unittest. Basically, we will need to change here https://github.com/dmlc/gluon-nlp/blob/20321598a732d254d4a9e036eecbe0ae8abb48a0/src/gluonnlp/attention_cell.py#L500-L501

File: src/operator/nn/softmax.cc [hunk 2]
```diff
@@ -187,5 +188,54 @@ NNVM_REGISTER_OP(_backward_softmax)
...
+.set_attr<nnvm::FListOutputNames>("FListInputNames",
+    [](const NodeAttrs& attrs){
+    return std::vector<std::string>{"data", "mask"};
+})
+.set_attr<nnvm::FListOutputNames>("FListOutputNames",
+    [](const NodeAttrs& attrs) {
+    return std::vector<std::string>{"output"};
+})
+.set_attr<FCompute>("FCompute<cpu>", MaskedSoftmaxCompute<cpu, mxnet_op::softmax_fwd>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseInOut{"_backward_masked_softmax"})
```
Comment (sxjscience) [inline src/operator/nn/softmax.cc:205]: We can try to register a `_zeros_like` for the mask.

See also https://github.com/apache/incubator-mxnet/blob/66488662ea6d33b8aeb64b632ea45194181f794a/src/operator/tensor/index_add_forward.cc#L113-L126

Comment (MoisesHer) [reply src/operator/nn/softmax.cc:205]: is this not equivalent to https://github.com/apache/incubator-mxnet/blob/95a8d41bf48c7716df361c117aeef65fb120c79c/src/operator/nn/softmax-inl.h#L1642-L1646 
?

Comment (sxjscience) [reply src/operator/nn/softmax.cc:205]: I'm not sure but I guess in the graph-pass, the zeros_like nodes may potentially be optimized.

File: src/operator/nn/softmax-inl.h [hunk 3]
```diff
@@ -907,6 +1632,56 @@ void SoftmaxGradCompute(const nnvm::NodeAttrs& attrs,
...
+                ctx.get_stream<xpu>(), out_ptr,
+                ograd_ptr, grad_data, mask_ptr,
+                inputs[1].shape_.get<ndim>(), inputs[2].shape_.get<ndim>(),
+                axis, static_cast<DType>(scale),
+                static_cast<DType>(temperature), ctx);
+          } else {
+            MaskedSoftmaxGrad<OP1, OP2, Req, negate, DType>(
+                ctx.get_stream<xpu>(), out_ptr,
+                ograd_ptr, grad_data, mask_ptr,
+                inputs[1].shape_.get<ndim>(), inputs[2].shape_.get<ndim>(),
```
Comment (sxjscience) [inline src/operator/nn/softmax-inl.h:1675]: `inputs[1]` is only used for providing the shape so it can be safely removed.

**Comment by MoisesHer** [comment]
@mxnet-bot run ci [edge]

**Comment by sxjscience** [comment]
LGTM

**Comment by grygielski** [comment]
@sxjscience Do you plan to change the current BERT model in Gluon-NLP repo to use this new op?

**Comment by sxjscience** [comment]
Yes, I’m waiting for it to appear in the nightly so I can use the new OP + run the benchmark again.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Pull request merged**